cpp/src/strings/strip.cu

```diff
@@ -0,0 +1,169 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strip.hpp>
+#include <cudf/utilities/error.hpp>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+#include <thrust/transform.h>
+#include <thrust/logical.h>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace
+{
+
+/**
+ * @brief Used as template parameter to divide size calculation from
+ * the actual string operation within a function.
+ *
+ * Useful when most of the logic is identical for both passes.
+ */
+enum TwoPass
+{
+    SizeOnly = 0, ///< calculate the size only
+    ExecuteOp     ///< run the string operation
+};
+
+/**
+ * @brief Strip characters from the beginning and/or end of a string.
+ *
+ * This functor strips the beginning and/or end of each string
+ * of any characters found in d_to_strip or whitespace if
+ * d_to_strip is empty.
+ *
+ * @tparam Pass Allows computing only the size of the output
+ *              or writing the output to device memory.
+ */
+template <TwoPass Pass=SizeOnly>
+struct strip_fn
+{
+    column_device_view const d_strings;
+    strip_type stype; // right, left, or both
+    string_view d_to_strip;
+    int32_t const* d_offsets{};
+    char* d_chars{};
+
+    __device__ size_type operator()( size_type idx )
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        string_view d_str = d_strings.element<string_view>(idx);
+        size_type length = d_str.length();
+        size_type left_offset = 0;
+        auto itr = d_str.begin();
+        if( stype==strip_type::LEFT || stype==strip_type::BOTH )
+        {
+            for( ; itr != d_str.end(); )
+            {
+                char_utf8 chr = *itr++;
+                if( d_to_strip.empty() ? (chr > ' ') : // whitespace
+                    thrust::none_of( thrust::seq, d_to_strip.begin(), d_to_strip.end(),
+                                        [chr] __device__ (char_utf8 c) {return c==chr;}) )
+                    break;
+                left_offset = itr.byte_offset();
+            }
+        }
+        size_type right_offset = d_str.size_bytes();
+        if( stype==strip_type::RIGHT || stype==strip_type::BOTH )
+        {
+            itr = d_str.end();
+            for( size_type n=0; n < length; ++n )
+            {
+                char_utf8 chr = *(--itr);
+                if( d_to_strip.empty() ? (chr > ' ') : // whitespace
+                    thrust::none_of( thrust::seq, d_to_strip.begin(), d_to_strip.end(),
+                        [chr] __device__ (char_utf8 c) {return c==chr;} ) )
+                    break;
```

Comment:
Ternary operators ? probably shouldn't be used inside an if statement.

Reply:
> Ternary operators ? probably shouldn't be used inside an if statement.

Why is that? The written out code looks less clean to me:
```
if( d_to_strip.empty() )
{
   if(chr > ' ')
      break;
}
else if( thrust::none_of( thrust::seq, d_to_strip.begin(), d_to_strip.end(),
                          [chr] __device__ (char_utf8 c) {return c==chr;} ) )
   break;
```
current code:
```
if( d_to_strip.empty() ? (chr > ' ') :
    thrust::none_of( thrust::seq, d_to_strip.begin(), d_to_strip.end(),
                     [chr] __device__ (char_utf8 c) {return c==chr;} ) )
   break;
```

Reply:
The purpose of the ternary operator is **assignment**, specifically conditional assignment.
The purpose of an if statement is to add **conditional logic**. 

When you use the ternary operator inside an if statement, you are effectively using conditional assignment inside a conditional statement, which for me at least is very confusing. Two alternative options that I think are cleaner are:

Note: My preference is **alternative 2**. It maintains the limited scope of the `boolean` to the `if` conditional and it can also be reused in both spots.

```cpp
// alternative 1
bool nothing_left_to_strip = d_to_strip.empty() ? chr > ' ' :
    thrust::none_of( thrust::seq, d_to_strip.begin(), d_to_strip.end(),
        [chr] __device__ (char_utf8 c) { return c == chr; })

if (nothing_left_to_strip)
    break;
```
and 
```cpp
// alternate 2 (refactor into function)
auto nothing_to_strip(string_view sv, char_utf8 c) -> bool {
    return sv.empty() ? c> ' ' :
        thrust::none_of(thrust::seq, sv.begin(), sv.end(),
            [c] __device__ (char_utf8 e) { return e == c; });
}

// and then
if (nothing_to_strip(d_to_strip, chr))
    break;
```